bse/driver-jack.cc

```diff
@@ -0,0 +1,905 @@
+can use helper threads to complete the job.
+
+Note that this optimization works best (no buffering/latency) if the engine
+block size and the jack block size are equal. It also works well if the jack
+block size is an integer multiple fo the engine block size.
+
+This avoids latency and buffering. However this may pose stricter RT
+requirements onto BEAST. So whether it runs as dropout-free as the current
+version would remain to be seen. A not so realtime version would buffer
+M complete blocks of N samples, still avoiding partially filled buffers.
+------------------------------------------------------------------------*/
+
+/**
+ * This function uses std::copy to copy the n_values of data from ivalues
+ * to ovalues. If a specialized version is available in bseblockutils,
+ * then this - usually faster - version will be used.
+ *
+ * The data in ivalues and ovalues may not overlap.
+ */
+template<class Data> void
+fast_copy (uint        n_values,
+           Data       *ovalues,
+	   const Data *ivalues)
+{
+  copy (ivalues, ivalues + n_values, ovalues);
+}
+
+template<> void
+fast_copy (uint         n_values,
+           float       *ovalues,
+           const float *ivalues)
+{
+  Block::copy (n_values, ovalues, ivalues);
+}
+
+#if 0 // <- avoid unused warning
+template<> void
+fast_copy (uint	         n_values,
+           uint32       *ovalues,
+           const uint32 *ivalues)
+{
+  Block::copy (n_values, ovalues, ivalues);
+}
+#endif
+
+/**
+ * The FrameRingBuffer class implements a ringbuffer for the communication
+ * between two threads. One thread - the producer thread - may only write
+ * data to the ringbuffer. The other thread - the consumer thread - may
+ * only read data from the ringbuffer.
+ *
+ * Given that these two threads only use the appropriate functions, no
+ * other synchronization is required to ensure that the data gets safely
+ * from the producer thread to the consumer thread. However, all operations
+ * that are provided by the ringbuffer are non-blocking, so that you may
+ * need a condition or other synchronization primitive if you want the
+ * producer and/or consumer to block if the ringbuffer is full/empty.
+ *
+ * Implementation: the synchronization between the two threads is only
+ * implemented by two index variables (read_frame_pos and write_frame_pos)
+ * for which atomic integer reads and writes are required. Since the
+ * producer thread only modifies the write_frame_pos and the consumer thread
+ * only modifies the read_frame_pos, no compare-and-swap or similar
+ * operations are needed to avoid concurrent writes.
+ */
+template<class T>
+class FrameRingBuffer {
+  //BIRNET_PRIVATE_COPY (FrameRingBuffer);
+private:
+  vector<vector<T> >  channel_buffer_;
+  std::atomic<int>    atomic_read_frame_pos_;
+  std::atomic<int>    atomic_write_frame_pos_;
+  uint                channel_buffer_size_;       // = n_frames + 1; the extra frame allows us to
```

Comment:
This is C++17, always and unconditionally initialize all struct members that don't have a ctor. If in doubt, just add =0, =false, =EnumType(0).

Reply:
Of course. Fixed. Seems the jack driver predates C++11.